akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolInterfaceActor.scala

```diff
@@ -174,6 +187,16 @@ private class PoolInterfaceActor(gateway: PoolGateway)(implicit fm: Materializer
         onCompleteThenStop()
       } else
         debug(s"Deferring shutting down host connection pool until all [$remainingRequested] responses have been dispatched")
+
+    case Prune ⇒ // Kick all expired requests out of the queue and fail them
+      while (inputBuffer.nonEmpty && inputBuffer.peek().deadline.filter(!_.hasTimeLeft()).isDefined) {
```

Comment:
Perhaps `!inputBuffer.peek().deadline.forall(_.hasTimeLeft())`?

Nice that this can just look at the side of the buffer that has the oldest elements 👍